v2/internal/resolver/resolver.go

```diff
@@ -177,14 +183,64 @@ func (r *Resolver) ResolveReference(ctx context.Context, ref genruntime.Namespac
 	return metaObj, nil
 }
 
+type ResolveOwnerResult string
+
+const (
+	// OwnerFoundKubernetes indicates the owner was found in Kubernetes.
+	OwnerFoundKubernetes = ResolveOwnerResult("OwnerFoundKubernetes")
+	// OwnerFoundARM indicates the owner is an ARM ID. The resource the ARM ID points to may or may not exist in Azure currently.
+	OwnerFoundARM = ResolveOwnerResult("OwnerFoundARM")
+	// OwnerNotExpected indicates that this resource is not expected to have any owner. (Example: ResourceGroup)
+	OwnerNotExpected = ResolveOwnerResult("OwnerNotExpected")
+)
+
+type OwnerDetails struct {
+	Result ResolveOwnerResult
+	Owner  genruntime.ARMMetaObject
+	ARMID  string
+}
+
+func (det OwnerDetails) FoundKuberentesOwner() bool {
+	if det.Result == OwnerNotExpected || det.Result == OwnerFoundARM {
+		// If no owner is expected or the owner is only in ARM, no need to assign ownership in Kubernetes
+		return false
+	}
+
+	return true
+}
+
+func OwnerDetailsFromKubernetes(owner genruntime.ARMMetaObject) OwnerDetails {
+	return OwnerDetails{
+		Result: OwnerFoundKubernetes,
+		Owner:  owner,
+	}
+}
+
+func OwnerDetailsFromARM(armID string) OwnerDetails {
+	return OwnerDetails{
+		Result: OwnerFoundARM,
+		ARMID:  armID,
+	}
+}
+
+func OwnerDetailsNotExpected() OwnerDetails {
+	return OwnerDetails{
+		Result: OwnerNotExpected,
+	}
+}
+
 // ResolveOwner returns the MetaObject for the given resource's owner. If the resource is supposed to have
 // an owner but doesn't, this returns an ReferenceNotFound error. If the resource is not supposed
-// to have an owner (for example, ResourceGroup), returns nil.
-func (r *Resolver) ResolveOwner(ctx context.Context, obj genruntime.ARMOwnedMetaObject) (genruntime.ARMMetaObject, error) {
+// to have an owner (for example, ResourceGroup) or the owner points to a raw ARM ID, returns nil.
+func (r *Resolver) ResolveOwner(ctx context.Context, obj genruntime.ARMOwnedMetaObject) (OwnerDetails, error) {
```

Comment:
Seems odd that `ResolveOwner` returns nil even if it has an owner, just because it's an ARM ID.

Maybe rename to `ResolveClusterOwner()` or `ResolveKubernetesOwner()`?

Reply:
Ah it doesn't actually return `nil` anymore. I refactored the method and added the `OwnerDetails` return type because I also didn't like this. Now it always has a return and that return has the `Result` set to either:
1. `OwnerFoundKubernetes` (if the owner was found in k8s)
2. `OwnerFoundARM` (if the owner was found in ARM)
3. `OwnerNotExpected` (if the owner was not found but also not expected)
4. an error (if an owner was not found but _was_ expected).

Reply:
I've fixed this comment.